### PR TITLE
More uniform handling of variable patterns in Btermdn.

### PR DESCRIPTION
**Pull request by ppedrot**
We first check that a pattern variable has a body before looking it up in the transparent state. This only affects discriminated nets. The behaviour or non-discriminated nets is very weird and asymmetrical since in patterns variables match everything but in terms they match only variables, regardless of the existence of a body.

### CODE REVIEW

**Comment by ppedrot** [comment]
Bench: https://gitlab.com/coq/coq/-/jobs/1465095402

**Comment by ppedrot** [comment]
Small example extracted from fiat-parsers:
```coq
Require Export Setoid.
Unset Implicit Arguments.

  Axiom t : Type -> Type.
  Axiom key : Type.

  Axiom fold : forall [elt A : Type], (key -> elt -> A -> A) -> t elt -> A -> A.

  Axiom map2 :
  forall [elt elt' elt'' : Type],
  (option elt -> option elt' -> option elt'') -> t elt -> t elt' -> t elt''.

  Definition lift_relation_gen_hetero {A B P}
             (and : P -> P -> P) (True : P)
             (R : A -> B -> P) (defaultA : A) (defaultB : B)
    : t A -> t B -> P
    := fun m1 m2
       => fold
            (fun _ => and)
            (map2
               (fun x1 x2
                => match x1, x2 with
                   | Some x1, Some x2 => Some (R x1 x2)
                   | Some x, None => Some (R x defaultB)
                   | None, Some x => Some (R defaultA x)
                   | None, None => None
                   end)
               m1 m2)
            True.

  Definition lift_relation_hetero {A B} := @lift_relation_gen_hetero A B Prop and True.


    Section rel.
    Context {P} {Q : P -> Prop} {and : P -> P -> P} {True' : P}
            {A} {default : A}.

    Local Coercion Q : P >-> Sortclass.

    Section with_R.
      Context {R : A -> A -> P}.
      Local Notation R' := (fun x y => Q (R x y)).
      Global Instance lift_relation_gen_hetero_Reflexive {HR : @Reflexive A R'} : Reflexive (@lift_relation_gen_hetero A A P and True' R default default) | 5.
      Proof. Admitted.

    End with_R.

  End rel.

  Module Params.
    Definition P := Prop.
    Definition Q : P -> Prop := fun H => H.
    Definition and : P -> P -> P := and.
    Definition True : P := True.

    Definition lift_relation_gen_hetero_Reflexive
      := @lift_relation_gen_hetero_Reflexive P Q and True.
    Global Existing Instances lift_relation_gen_hetero_Reflexive | 5.
  End Params.

  Global Instance lift_relation_hetero_Reflexive {A R} {HR : @Reflexive A R} {default}
    : Reflexive (lift_relation_hetero R default default) | 5 := _.
```

**Comment by ppedrot** [comment]
Even simpler:
```
Class Reflexive0 {A : Type} (R : A -> Prop) : Prop := {}.

Definition foo {A R} {HR : @Reflexive0 A R} : @Reflexive0 A (fun x => R x) := _.
```
The problem is that making the variable opaque prevents syntactic matching up to η-expansion, even though unification would succeed on this. I don't know if this semantics is intended, but it departs from the one for constants where a similar example fails on master.
```
Class Reflexive0 {A : Type} (R : A -> Prop) : Prop := {}.

Axiom A : Type.
Axiom R : A -> Prop.
Local Declare Instance HR : @Reflexive0 A R.

Definition foo : @Reflexive0 A (fun x => R x) := _.
```

cc @mattam82

**Comment by ppedrot** [comment]
@coqbot ci minimize

**Comment by ppedrot** [comment]
As mentioned on Zulip, in 8.12 constants were handled the same way as variables w.r.t. η-expansion. This is not the case in 8.13, and the commit that introduced the change was:
```
commit 4ebccb9d2722a7323cb7fce5c9e00bf2eea5b69f
Author: Pierre-Marie Pédrot <pierre-marie.pedrot@inria.fr>
Date:   Mon Jun 22 12:59:06 2020 +0200

    Dnets now consider axioms as being opaque for pattern recognition.
```
I am a bit befuzzled that there is no test that caught it, nor any dev in the CI. The fact that we get stability by η-expansion out of this somewhat unrelated change is dubious to say the least, and should be handled in a principle way rather than by unexpected side-effects of dnet matching.

**Comment by ppedrot** [comment]
@coqbot ci minimize

**Comment by JasonGross** [comment]
@ppedrot Do you still have access to 0e84f003ce65f7d9f45a5823c8527735685dfb41?  If so, could you force-push that commit to JasonGross/coq:test-coqbot-minimize (https://github.com/coq/coq/pull/14579).  I'm not sure how to access orphan commits, and I'd like to debug the failure to fully minimize from ci-color and ci-perennial (in addition to the issues with ci-fiat_crypto)

**Comment by ppedrot** [comment]
@JasonGross done

**Comment by ppedrot** [comment]
@coqbot ci minimize
```
ci-fiat_crypto
```

**Comment by JasonGross** [comment]
@ppedrot the request needs to have the CI's listed on the same line as the coqbot tag, unquoted

**Comment by ppedrot** [comment]
Minimal example from fiat-crypto:
```
Module map.
  Class map {value} := mk {
    rep : Type;
    get: rep -> option value;
    empty : rep;
    put : rep -> value -> rep;
  }.
  Arguments map : clear implicits.
  Class ok {value : Type} {map : map value}: Prop := {
    get_put_diff : forall m v, get (put m v) = None;
  }.
  Arguments ok {_} _.

End map.

Axiom wordX : Set.
Axiom e : wordX.

Section Equiv.

Context {Registers: map.map wordX}
        {mem: map.map Coq.Init.Byte.byte}.

Context (Registers_ok: @map.ok wordX Registers)
        (mem_ok: @map.ok Byte.byte mem).

Goal map.get (map.put map.empty e) = None.
Proof.
clear -Registers_ok.
Set Typeclasses Debug.
apply map.get_put_diff.
Abort.
```

**Comment by ppedrot** [comment]
Bench: https://gitlab.com/coq/coq/-/jobs/1479760591

**Comment by ppedrot** [comment]
```
┌─────────────────────────────┬─────────────────────────┬───────────────────────────────────────┬────────────────────────────────────────┬─────────────────────────┬─────────────────┐
│                             │      user time [s]      │              CPU cycles               │            CPU instructions            │  max resident mem [KB]  │   mem faults    │
│                             │                         │                                       │                                        │                         │                 │
│                package_name │     NEW      OLD  PDIFF │            NEW             OLD  PDIFF │            NEW             OLD   PDIFF │     NEW      OLD  PDIFF │ NEW  OLD  PDIFF │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│  coq-performance-tests-lite │ 1400.66  1552.08  -9.76 │  3911772060508   4324198148479  -9.54 │  6330635099304   7111758225574  -10.98 │ 2085060  2085116  -0.00 │   0    0    nan │
│       coq-engine-bench-lite │  290.49   292.66  -0.74 │   818595655805    824521756725  -0.72 │  1496071713736   1497429602353   -0.09 │ 1412000  1411828   0.01 │   0    0    nan │
│                   coq-flocq │  145.92   146.47  -0.38 │   404946095198    405636876760  -0.17 │   503778294546    503613404134    0.03 │  810352   810412  -0.01 │   0    0    nan │
│                  coq-stdlib │  411.41   412.88  -0.36 │  1152249688794   1153766456331  -0.13 │  1467296222395   1466982360227    0.02 │  611908   610160   0.29 │   0    0    nan │
│                 coq-bignums │   55.42    55.60  -0.32 │   153771819631    153227838992   0.36 │   206859448348    206845170454    0.01 │  462584   463092  -0.11 │   0    0    nan │
│                    coq-hott │  296.58   296.87  -0.10 │   815541509359    816139781579  -0.07 │  1267188362129   1267297626665   -0.01 │  570916   570852   0.01 │   0    0    nan │
│                coq-coqprime │  326.85   327.16  -0.09 │   908894977387    909799227083  -0.10 │  1595648296994   1594835359992    0.05 │  916468   916496  -0.00 │   0    0    nan │
│                  coq-geocoq │ 1317.98  1318.69  -0.05 │  3673292130944   3676645210973  -0.09 │  5419681747635   5417306653647    0.04 │  984356  1070256  -8.03 │   0    0    nan │
│                coq-bedrock2 │  302.64   302.75  -0.04 │   838549680011    838784902523  -0.03 │  1492847647006   1491999001648    0.06 │  799904   799968  -0.01 │   0    0    nan │
│            coq-math-classes │  172.24   172.27  -0.02 │   479713337495    480194545797  -0.10 │   640415285824    640934103712   -0.08 │  482088   481856   0.05 │   0    0    nan │
│              coq-coquelicot │   68.42    68.36   0.09 │   187295102396    186319247611   0.52 │   240719708868    239757442682    0.40 │  558232   558408  -0.03 │   0    0    nan │
│                   coq-color │  482.79   482.22   0.12 │  1347743468550   1342634229411   0.38 │  1697966339841   1695843497296    0.13 │ 1156380  1153684   0.23 │   0    0    nan │
│               coq-perennial │ 7174.06  7165.20   0.12 │ 19982616742369  19961846492804   0.10 │ 30633013177272  30642547274315   -0.03 │ 2989252  3143120  -4.90 │   0    0    nan │
│             coq-fiat-crypto │ 4802.50  4787.11   0.32 │ 13339973908593  13297510415957   0.32 │ 23556284282590  23554276515861    0.01 │ 2222348  2221808   0.02 │   0    0    nan │
│               coq-fiat-core │  107.93   107.53   0.37 │   304889069074    303400255937   0.49 │   411614633213    410071875991    0.38 │  481228   480964   0.05 │   0    0    nan │
│                coq-rewriter │  684.78   681.99   0.41 │  1898925590445   1893582019348   0.28 │  2930763330488   2928586905271    0.07 │ 1073932  1073848   0.01 │   0    0    nan │
│               coq-fourcolor │ 2556.18  2545.31   0.43 │  7115235753853   7097040235254   0.26 │ 12236472020547  12238122130767   -0.01 │  732772   732824  -0.01 │   0    0    nan │
│      coq-mathcomp-ssreflect │   54.15    53.91   0.45 │   147494919845    146901408924   0.40 │   190922095399    190953477155   -0.02 │  527620   527560   0.01 │   0    0    nan │
│             coq-lambda-rust │ 1266.43  1260.78   0.45 │  3523770357428   3507478771300   0.46 │  5084887724263   5080103769902    0.09 │ 1107424  1097856   0.87 │   0    0    nan │
│       coq-mathcomp-fingroup │   48.25    48.00   0.52 │   133327532822    132513794864   0.61 │   186316308744    186347045010   -0.02 │  479624   479640  -0.00 │   0    0    nan │
│      coq-mathcomp-odd-order │ 1044.12  1038.24   0.57 │  2906299843018   2889701445687   0.57 │  4619850827214   4619716889845    0.00 │  942448   942432   0.00 │   0    0    nan │
│          coq-mathcomp-field │  221.61   220.34   0.58 │   616597999096    613492712215   0.51 │   938347253896    938347985051   -0.00 │  664308   663576   0.11 │   0    0    nan │
│ coq-rewriter-perf-SuperFast │ 1630.83  1619.20   0.72 │  4540523077938   4505412061643   0.78 │  7151345212403   7154543086826   -0.04 │ 1073856  1073804   0.00 │   0    0    nan │
│                 coq-unimath │ 4662.84  4629.19   0.73 │ 12981242758619  12881566015129   0.77 │ 24872877464515  24875680865749   -0.01 │ 3713352  3711896   0.04 │   0    0    nan │
│              coq-verdi-raft │ 1101.16  1092.54   0.79 │  3062016113895   3042504790017   0.64 │  4304631640448   4302828604326    0.04 │  916400   920580  -0.45 │   0    0    nan │
│       coq-mathcomp-solvable │  169.46   168.08   0.82 │   470959103311    467360152942   0.77 │   665504648825    665516316583   -0.00 │  694648   694848  -0.03 │   0    0    nan │
│        coq-mathcomp-algebra │  140.97   139.80   0.84 │   391132331733    389091675034   0.52 │   509694296844    509650577510    0.01 │  556420   556376   0.01 │   0    0    nan │
│                   coq-verdi │   98.74    97.91   0.85 │   271215790068    269293238917   0.71 │   379495935931    378578642984    0.24 │  547316   547004   0.06 │   0    0    nan │
│                    coq-corn │ 1461.81  1449.30   0.86 │  4071257404737   4039997434013   0.77 │  5864317793401   5862201337883    0.04 │  807096   807320  -0.03 │   0    0    nan │
│                    coq-core │  108.05   107.11   0.88 │   298300646582    297810882339   0.16 │   461274475854    461338106958   -0.01 │  295896   295908  -0.00 │   0    0    nan │
│                 coq-coqutil │   59.86    59.33   0.89 │   166208506086    164821347839   0.84 │   218419499882    217869096885    0.25 │  601928   602084  -0.03 │   0    0    nan │
│            coq-fiat-parsers │  583.55   578.23   0.92 │  1631184266494   1620061047270   0.69 │  2488893547096   2485278617664    0.15 │ 3485496  3485180   0.01 │   0    0    nan │
│      coq-mathcomp-character │  149.75   148.34   0.95 │   415585482607    412882182002   0.65 │   583121806751    583159551090   -0.01 │  732608   732540   0.01 │   0    0    nan │
└─────────────────────────────┴─────────────────────────┴───────────────────────────────────────┴────────────────────────────────────────┴─────────────────────────┴─────────────────┘
```
Pretty good.

**Comment by ppedrot** [comment]
@JasonGross this divides by about two the compilation time of all the instances of `typeclass_reification` from `performance-tests`. I am a bit surprised that is has no other observable effects in actual developments but maybe we don't stress typeclasses so much there.

**Comment by Janno** [comment]
This could be very interesting for the development I am working on but I don't think I'll ever be able to test that against coq/master. We often have local variables (without bodies) appearing in patterns. I suppose this change should be expected to speed things up by avoiding the transparent state lookup?

**Comment by JasonGross** [comment]
Wow, that's amazing @ppedrot !  Does it have something to do with how many binders the instances are under?  I wonder if there are any other low-hanging fruit for typeclasses

**Comment by ppedrot** [comment]
>  I suppose this change should be expected to speed things up by avoiding the transparent state lookup?
> Does it have something to do with how many binders the instances are under?

Nah, it simply prunes some typeclass instances early, preventing them to be checked for unification with the goal (which might be very expensive). The main difference now is that variables without a body are considered rigid for pattern recognition (i.e. they only match themselves, where they would match anything before)/

**Comment by JasonGross** [comment]
> preventing them to be checked for unification with the goal (which might be very expensive). 

My hypothesis was that this unification is expensive because it's expensive to allocate evars under many binders, and IIUC evars must be allocated for all of the dependent premises of the instance in order to perform unification.  Are you saying that the cost lies in some other part of unification?

**Comment by ppedrot** [comment]
> Are you saying that the cost lies in some other part of unification?

No, I am not saying anything about unification beyond the fact it can be expensive (e.g. we have quadratic behaviours to compute term groundness, but there are many other ones, and evars are part of the big picture). My point is just that this PR simply short-circuits unification which is now not called at all in some easy-to-recognize cases.

**Comment by ppedrot** [comment]
This is ready and needs an assignee.

File: tactics/btermdn.ml [hunk 1]
```diff
@@ -130,9 +131,8 @@ let constr_pat_discr env ts t =
   | PVar v, args ->
     begin match ts with
     | None -> None
```
Comment (SkySkimmer) [inline tactics/btermdn.ml:133]: Why is `PVar` different from `PRef VarRef` when ts=None?
Why does ts=None give false in evaluable_named but true in evaluable_constant?
Are these questions related to the new FIXME?

File: test-suite/bugs/closed/bug_14722.v [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        {mem: map.map Coq.Init.Byte.byte}.
+
+Context (Registers_ok: @map.ok wordX Registers)
+        (mem_ok: @map.ok Byte.byte mem).
+
+(** Check that cleared variables are correctly handled *)
+Goal map.get (map.put map.empty e) = None.
+Proof.
+clear -Registers_ok.
+Set Typeclasses Debug.
```
Comment (SkySkimmer) [inline test-suite/bugs/closed/bug_14722.v:31]: Why do we set debug here?

**Review by SkySkimmer — APPROVED** [review]

File: tactics/btermdn.ml (hunk 1 above)
Comment (ppedrot) [reply tactics/btermdn.ml:133]: The `PRef VarRef` case is essentially dead code, there might be a way to trigger it with handwritten patterns in hints but even there it's not clear. For the rest, see #14753.

File: test-suite/bugs/closed/bug_14722.v (hunk 2 above)
Comment (ppedrot) [reply test-suite/bugs/closed/bug_14722.v:31]: Remnant of my debug session. I'll remove it, it's useless there.

Comment (ppedrot) [reply test-suite/bugs/closed/bug_14722.v]: Done.

**Comment by SkySkimmer** [comment]
@coqbot merge now
